### PR TITLE
Locking literals from being freed until global termination

### LINKED ISSUE #1764: Freeing String literals
Right now, calling `free` on a string literal will cause a segfault.

There should be a check in `free` to check whether the current object is a literal, and if so, not do anything. The `String` cleanup should be the only one which can free the string literals, perhaps via setting a "secret" flag in the `String` instance.

This will allow us to do stuff like `"literal" & variable` without problems. (Right now `"literal" << variable` must be used to not leak).

### PR DESCRIPTION
**Pull request by davidpiuva**
Alternative solution to https://github.com/magic-lang/ooc-kean/pull/1776

Fixes #1764

### CODE REVIEW

**Comment by marcusnaslund** [comment]
This is what I had in mind, actually.

Another alternative is to leave `free` as-is and have the concatenation operators `<<`, `>>`, `&` only call free on the argument(s) when they're not literals. Not sure that's any better though. @sebastianbaginski ?

**Comment by davidpiuva** [comment]
I am still getting leaks when running this test:
```
result: String
for (i in 1 .. 100) {
	result = "ab" & "cd"
	expect(result, is equal to("abcd"))
}
```
Maybe something wrong with my global freeing.

**Comment by sebastianbaginski** [comment]
I'd like to avoid having different interfaces for literals and regular strings. Actually I like this solution.

File: source/system/string/string_literal.c [hunk 1]
```diff
@@ -59,8 +59,10 @@ void string_literal_free(void* ptr) {
 void string_literal_free_all() {
 	if (_literals) {
 		for (size_t i=0; i<_literalsCount; ++i)
-			if (_literals[i])
+			if (_literals[i]) {
+				String__String__unlock(_literals[i]);
```
Comment (davidpiuva) [inline source/system/string/string_literal.c]: I don't know how to access the variable directly with mangled names so I made a setter method.

**Comment by davidpiuva** [comment]
When I added debug printing, I got crashed instead of leaks when it suddenly started the test by freeing all literals globally.

**Comment by marcusnaslund** [comment]
@davidpiuva Your code example should leak, since you're not freeing `result`.

**Comment by davidpiuva** [comment]
That makes sense

**Comment by davidpiuva** [comment]
I probably got a crash from debug printing because the message itself to print had already been freed when reaching the point to debug.

**Comment by davidpiuva** [comment]
@sebastianbaginski Peer review?

File: source/system/string/string_literal.c (hunk 1 above)
Comment (sebastianbaginski) [inline source/system/string/string_literal.c]: Don't bother, this is ok.

**Comment by sebastianbaginski** [comment]
Nicely done.

File: test/system/StringTest.ooc [hunk 2]
```diff
@@ -11,6 +11,15 @@ use unit
 StringTest: class extends Fixture {
 	init: func {
 		super("String")
+		this add("literal", func {
+			// The & operand tries to free the literals but if the lock works then they should resist the attempt
+			result: String
+			for (i in 1 .. 100) {
+				result = "ab" & "cd"
```
Comment (marcusnaslund) [inline test/system/StringTest.ooc]: For even better effect you could declare `ab = "ab"` and `cd = "cd"` outside the loop and use them.

File: source/system/String.ooc [hunk 3]
```diff
@@ -16,15 +16,18 @@ String: class {
 	// Avoid direct buffer access, as it breaks immutability.
 	_buffer: CharBuffer
 	size ::= this _buffer size
-
+	_locked := false // Only global termination may unlock and free literals
+	_unlock: func { this _locked = false }
```
Comment (marcusnaslund) [inline source/system/String.ooc]: No methods above `init`, put it somewhere else. Right below `free` is OK.

File: test/system/StringTest.ooc (hunk 2 above)
Comment (marcusnaslund) [inline test/system/StringTest.ooc]: Or both.

**Comment by marcusnaslund** [comment]
That's all I've got. Well done.

File: test/system/StringTest.ooc [hunk 4]
```diff
@@ -11,6 +11,22 @@ use unit
...
+				expect(result, is equal to("abcd"))
+				result free()
+			}
+			ab := "ab"
```
Comment (davidpiuva) [inline test/system/StringTest.ooc]: Updated

**Pull request merged**